File: src/lib/agents/utils/dedup.ts
import { calcDistanceM } from "@/lib/llm";
import type { RawCrawledPlace } from "../state";

import { normalizeName } from "@/lib/normalize";
// Re-export for backward compatibility
export { normalizeName };

/**
 * Check if two places are the same:
 * - normalized names match AND
 * - if both have coordinates, within 200m
 */
export function isSamePlace(a: RawCrawledPlace, b: RawCrawledPlace): boolean {
  const nameA = normalizeName(a.name);
  const nameB = normalizeName(b.name);

  if (nameA !== nameB) return false;

  // If both have coords, check distance
  if (
    a.lat != null &&
    a.lng != null &&
    b.lat != null &&
    b.lng != null
  ) {
    const dist = calcDistanceM(a.lat, a.lng, b.lat, b.lng);
    return dist <= 200;
  }

  // Same name, no coords to compare — treat as same
  return true;
}

interface MergedPlace extends RawCrawledPlace {
  sources: { source: string; sourceUrl?: string; rating?: number; reviewCount?: number; snippet?: string; metadata?: string }[];
}

/**
 * Deduplicate crawled places:
 * - group by normalized name + proximity
 * - merge sources
 * - pick the most complete data as the base
 */
export function deduplicatePlaces(places: RawCrawledPlace[]): MergedPlace[] {
  const groups: MergedPlace[] = [];

  for (const place of places) {
    const existing = groups.find((g) => isSamePlace(g, place));

    if (existing) {
      // Merge source info
      const alreadyHasSource = existing.sources.some(
        (s) => s.source === place.source
      );
      if (!alreadyHasSource) {
        existing.sources.push({
          source: place.source,
          sourceUrl: place.sourceUrl,
          rating: place.rating,
          reviewCount: place.reviewCount,
          snippet: place.snippet,
          metadata: place.metadata,
        });
      }

      // Fill in missing fields from the new place
      if (!existing.lat && place.lat) existing.lat = place.lat;
      if (!existing.lng && place.lng) existing.lng = place.lng;
      if (!existing.address && place.address) existing.address = place.address;
      if (!existing.category && place.category) existing.category = place.category;
      if (!existing.rating && place.rating) existing.rating = place.rating;
      if (!existing.tags && place.tags) existing.tags = place.tags;
    } else {
      groups.push({
        ...place,
        sources: [
          {
            source: place.source,
            sourceUrl: place.sourceUrl,
            rating: place.rating,
            reviewCount: place.reviewCount,
            snippet: place.snippet,
            metadata: place.metadata,
          },
        ],
      });
    }
  }

  return groups;
}

File: docs/samples/agents/dedup.py
"""Deduplication utilities for crawled places."""

import re

from ..llm_service import calc_distance_m


def normalize_name(name: str) -> str:
    """Normalize place name for dedup matching."""
    result = name.strip().lower()
    result = re.sub(r"\s*(본점|지점|점|역점|직영점)\s*$", "", result)
    result = re.sub(r"\s+", " ", result)
    return result


def is_same_place(a: dict, b: dict) -> bool:
    """Check if two places are the same by name + proximity."""
    if normalize_name(a["name"]) != normalize_name(b["name"]):
        return False

    if (
        a.get("lat") is not None
        and a.get("lng") is not None
        and b.get("lat") is not None
        and b.get("lng") is not None
    ):
        dist = calc_distance_m(a["lat"], a["lng"], b["lat"], b["lng"])
        return dist <= 200

    return True


def deduplicate_places(places: list[dict]) -> list[dict]:
    """Deduplicate crawled places by name + proximity, merge sources."""
    groups: list[dict] = []

    for place in places:
        existing = None
        for g in groups:
            if is_same_place(g, place):
                existing = g
                break

        if existing:
            # Merge source info
            already_has = any(s["source"] == place.get("source") for s in existing["sources"])
            if not already_has:
                existing["sources"].append({
                    "source": place.get("source", ""),
                    "sourceUrl": place.get("sourceUrl"),
                    "rating": place.get("rating"),
                    "reviewCount": place.get("reviewCount"),
                    "snippet": place.get("snippet"),
                    "metadata": place.get("metadata"),
                })

            # Fill in missing fields
            if not existing.get("lat") and place.get("lat"):
                existing["lat"] = place["lat"]
            if not existing.get("lng") and place.get("lng"):
                existing["lng"] = place["lng"]
            if not existing.get("address") and place.get("address"):
                existing["address"] = place["address"]
            if not existing.get("category") and place.get("category"):
                existing["category"] = place["category"]
            if not existing.get("rating") and place.get("rating"):
                existing["rating"] = place["rating"]
            if not existing.get("tags") and place.get("tags"):
                existing["tags"] = place["tags"]
        else:
            groups.append({
                **place,
                "sources": [
                    {
                        "source": place.get("source", ""),
                        "sourceUrl": place.get("sourceUrl"),
                        "rating": place.get("rating"),
                        "reviewCount": place.get("reviewCount"),
                        "snippet": place.get("snippet"),
                        "metadata": place.get("metadata"),
                    }
                ],
            })

    return groups
